### PR TITLE
[subgroups][non_uniform_broadcast] Fix broadcasting index generation

### PR DESCRIPTION
**Pull request by Nuullll**
Dynamically activate half of the work items in the current subgroup instead of hardcoding the number of active work items.

Otherwise, if subgroup size == `NR_OF_ACTIVE_WORK_ITEMS` == 4, then we will encounter "divide-by-zero" error when evaluating `bcast_index % (n - NR_OF_ACTIVE_WORK_ITEMS)`.

Signed-off-by: Yilong Guo <yilong.guo@intel.com>

### CODE REVIEW

**Comment by Nuullll** [comment]
gentle ping :-)

**Comment by Nuullll** [comment]
@StuartDBrady Can you please take a look?

**Comment by Nuullll** [comment]
ping @svenvh

File: test_conformance/subgroups/subgroup_common_templates.h [hunk 1]
```diff
@@ -82,14 +82,21 @@ template <typename Ty, SubgroupsBroadcastOp operation> struct BC
...
+                    // reduce broadcasting index in case of non_uniform and
                     // last workgroup last subgroup
                     if (last_subgroup_size && j == nj - 1
                         && last_subgroup_size < NR_OF_ACTIVE_WORK_ITEMS)
                     {
                         bcast_if = bcast_index % last_subgroup_size;
                         bcast_elseif = bcast_if;
                     }
+                    // reduce broadcasting index in case subgroup size <=
+                    // NR_OF_ACTIVE_WORK_ITEMS (i.e. all items are active)
```
Comment (svenvh) [inline test_conformance/subgroups/subgroup_common_templates.h:94]: This highlights a problem with the broadcast tests: they currently are not meant to test subgroup sizes `<= NR_OF_ACTIVE_WORK_ITEMS` (i.e., 4).  Your proposed fix causes the test to skip an important aspect of non-uniform subgroup operations, as the `else` in `sub_group_non_uniform_broadcast_source` will not be executed.  That means the subgroup operation will not be tested properly when the subgroup size is `<= 4`, so I don't think we should commit this.

Instead, we should probably try to get rid of `NR_OF_ACTIVE_WORK_ITEMS` and use work-item masks to introduce divergence in the broadcast tests (as done for e.g. `sub_group_non_uniform_any`).

**Review by svenvh — CHANGES_REQUESTED** [review]

File: test_conformance/subgroups/subgroup_common_templates.h (hunk 1 above)
Comment (Nuullll) [reply test_conformance/subgroups/subgroup_common_templates.h:94]: Thanks for reviewing. I agree. Will update the PR.

**Comment by Nuullll** [comment]
> @Nuullll, would you be able to use the same mechanism as used for `sub_group_non_uniform_any`, instead? Without that, I don't think we test with all items active, for example.

@StuartDBrady Currently we have both "positive" (`if`) and "negative" (`else`) coverage for `broadcast` functions (which is good, because this pattern did expose our OpenCL implementation bugs):
```opencl
    if (xy[gid].x < (get_sub_group_size() >> 1)) {
        out[gid] = sub_group_non_uniform_broadcast(x, xy[gid].z);
    } else {
        out[gid] = sub_group_non_uniform_broadcast(x, xy[gid].w);
    }
```
Is it necessary to test the all-active situation? which would leave the `else` part to return undefined results.

Furthermore, I did try to use the `uint4 work_item_mask_vector` parameter to represent the active items. The problem is that we need to provide a corresponding broadcasting index for each mask, as the second parameter of `sub_group_non_uniform_broadcast` call. However, the current `gen()` implementation is independent of the actual mask:
https://github.com/KhronosGroup/OpenCL-CTS/blob/c73d6a341bdcc1c05fa995390672e56a4cf66c6c/test_conformance/subgroups/subhelpers.h#L1614-L1633
Of course, we can do the following (pseudo code):
```opencl
__kernel void test(..., uint4 work_item_mask_vector) {
    ...
    if (elect_work_item & work_item_mask)
        out[gid] = sub_group_non_uniform_broadcast(x, get_index_of_one_bit(work_item_mask_vector));
    else
        out[gid] = sub_group_non_uniform_broadcast(x, get_index_of_one_bit(~work_item_mask_vector));
```
`get_index_of_one_bit` could be anything as long as it produces an index of a set bit according to the dynamic mask value. And we need to implement the same `get_index_of_one_bit` algorithm in host `chk()` function. I think it's just another form of hardcoding on the broadcasting index.

In my opinion, the current approach (hardcoding the divergence condition both in kernel code and `gen()` implementation, and taking the broadcast index from the random input) is much simpler. Or I could have missed something, could you please give some advice?

**Comment by Nuullll** [comment]
any comments?

File: test_conformance/subgroups/subgroup_common_templates.h [hunk 2]
```diff
@@ -78,24 +78,14 @@ template <typename Ty, SubgroupsBroadcastOp operation> struct BC
                 int bcast_elseif = 0;
                 int bcast_index = (int)(genrand_int32(gMTdata) & 0x7fffffff)
                     % (d > n ? n : d);
+                int num_of_active_items = n >> 1;
```
[the remark this answers is not in the document]
Comment (lakshmih) [reply test_conformance/subgroups/subgroup_common_templates.h:81]: Will be good to add this check.

[the remark this answers is not in the document]
Comment (Nuullll) [reply test_conformance/subgroups/subgroup_common_templates.h:81]: thank you. updated.

[the remark this answers is not in the document]
Comment (Nuullll) [reply test_conformance/subgroups/subgroup_common_templates.h:81]: thanks. added checks

**Comment by Nuullll** [comment]
friendly ping.

**Review by svenvh — APPROVED** [review]

**Review by lakshmih — APPROVED** [review]

**Comment by bashbaug** [comment]
Merging as discussed in the March 12th teleconference.

**Pull request merged**